File: src/minimum_absolute_difference_between_elements_with_constraint_2817/solution_two_heaps.py
import heapq
from typing import List


def min_diff(nums: List[int], x: int) -> int:
    sorted_list = sorted((val, idx) for idx, val in enumerate(nums))

    min_heap = []
    max_heap = []
    result = 1e10

    for val, idx in sorted_list:
        heapq.heappush(min_heap, (idx, val))
        heapq.heappush(max_heap, (-idx, val))

        while len(min_heap) > 0 and min_heap[0][0] <= idx - x:
            result = min(result, val - heapq.heappop(min_heap)[1])
        while len(max_heap) > 0 and max_heap[0][0] <= -idx - x:
            result = min(result, val - heapq.heappop(max_heap)[1])

    return result


def test():
    print(min_diff([4, 3, 2, 4], 2))
    print(min_diff([5, 3, 2, 10, 15], 1))
    print(min_diff([1, 2, 3, 4], 3))
    print(min_diff([14, 111, 16], 1))

File: src/minimum_absolute_difference_between_elements_with_constraint_2817/solution_two_heaps.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

pub struct Solution;

impl Solution {
    fn min_absolute_difference(nums: Vec<i32>, x: i32) -> i32 {
        let mut sorted_list: Vec<(i32, usize)> = nums
            .into_iter()
            .enumerate()
            .map(|(idx, val)| (val, idx))
            .collect();
        sorted_list.sort();

        let x = x as usize;
        let mut min_heap: BinaryHeap<Reverse<(usize, i32)>> = BinaryHeap::new();
        let mut max_heap: BinaryHeap<(usize, i32)> = BinaryHeap::new();

        let mut result = i32::MAX;

        for &(val, idx) in &sorted_list {
            min_heap.push(Reverse((idx, val)));
            max_heap.push((idx, val));

            while let Some(&Reverse((heap_idx, heap_val))) = min_heap.peek() {
                if idx >= x && heap_idx <= idx - x {
                    result = result.min((val - heap_val).abs());
                    min_heap.pop();
                } else {
                    break;
                }
            }

            while let Some(&(heap_idx, heap_val)) = max_heap.peek() {
                if heap_idx >= idx + x {
                    result = result.min((val - heap_val).abs());
                    max_heap.pop();
                } else {
                    break;
                }
            }
        }

        result
    }
}

pub fn test() {
    println!("{}", Solution::min_absolute_difference(vec![4, 3, 2, 4], 2));
    println!(
        "{}",
        Solution::min_absolute_difference(vec![5, 3, 2, 10, 15], 1)
    );
    println!("{}", Solution::min_absolute_difference(vec![1, 2, 3, 4], 3));
    println!(
        "{}",
        Solution::min_absolute_difference(vec![14, 111, 16], 1)
    );
    println!("{}", Solution::min_absolute_difference(vec![71, 4], 1));
}
